biorxiv/polka_subset_experiment/04_regression_results_post_covid_preprints.py

```diff
@@ -0,0 +1,302 @@
+    "cosine",
+)
+print(document_pca_sim.shape)
+document_pca_sim
+
+document_to_pca_map = {
+    document: document_pca_sim[idx, :]
+    for idx, document in enumerate(biorxiv_vec_df.document.tolist())
+}
+
+biorxiv_pca_sim_df = (
+    pd.DataFrame.from_dict(document_to_pca_map, orient="index")
+    .rename(index=str, columns={col: f"pc{col+1}" for col in range(int(300))})
+    .reset_index()
+    .rename(index=str, columns={"index": "document"})
+    .assign(label="biorxiv")
+)
+biorxiv_pca_sim_df.head()
+
+# ## PC Regression
+
+# ### Logistic Regression
+
+# Goal here is to determine if we can figure out which PCs separate the bioRxiv subset from Polka et al.'s subset. Given that their dataset is only 60 papers we downsampled our dataset to contain only 60 papers.
+
+dataset_df = biorxiv_pca_sim_df.append(polka_pca_sim_df)
+dataset_df.head()
+
+model = LogisticRegressionCV(
+    cv=10, Cs=100, max_iter=1000, penalty="l1", solver="liblinear"
+)
+model.fit(
+    StandardScaler().fit_transform(dataset_df[[f"pc{idx+1}" for idx in range(50)]]),
+    dataset_df["label"],
+)
+
+best_result = list(filter(lambda x: x[1] == model.C_, enumerate(model.Cs_)))[0]
+print(best_result)
+
+print("Best CV Fold")
+print(model.scores_["polka"][:, best_result[0]])
+model.scores_["polka"][:, best_result[0]].mean()
+
+model_weights_df = pd.DataFrame.from_dict(
+    {
+        "weight": model.coef_[0],
+        "pc": list(range(1, 51)),
+    }
+)
+model_weights_df["pc"] = pd.Categorical(model_weights_df["pc"])
+model_weights_df.head()
+
+g = (
+    p9.ggplot(model_weights_df, p9.aes(x="pc", y="weight"))
+    + p9.geom_col(position=p9.position_dodge(width=5), fill="#253494")
+    + p9.coord_flip()
+    + p9.scale_x_discrete(limits=list(sorted(range(1, 51), reverse=True)))
+    + p9.theme_seaborn(context="paper", style="ticks", font_scale=1.1, font="Arial")
+    + p9.theme(figure_size=(10, 8))
+    + p9.labs(
+        title="Regression Model Weights", x="Princpial Component", y="Model Weight"
+    )
+)
+# g.save("output/figures/pca_log_regression_weights.svg")
+# g.save("output/figures/pca_log_regression_weights.png", dpi=250)
+print(g)
+
+fold_features = model.coefs_paths_["polka"].transpose(1, 0, 2)
+model_performance_df = pd.DataFrame.from_dict(
+    {
+        "feat_num": ((fold_features.astype(bool).sum(axis=1)) > 0).sum(axis=1),
+        "C": model.Cs_,
+        "score": model.scores_["polka"].mean(axis=0),
+    }
+)
+model_performance_df.head()
+
+# +
+fig, ax1 = plt.subplots()
+ax1.set_xscale("log")
+ax2 = plt.twinx()
+
+ax1.plot(
+    model_performance_df.C.tolist(),
+    model_performance_df.feat_num.tolist(),
+    label="Features",
+    marker=".",
+)
+ax1.set_ylabel("# of Features")
+ax1.set_xlabel("Inverse Regularization (C)")
+ax1.legend(loc=0)
+
+ax2.plot(
+    model_performance_df.C.tolist(),
+    model_performance_df.score.tolist(),
+    label="Score",
+    marker=".",
+    color="green",
+)
+ax2.set_ylabel("Score (Accuracy %)")
```

Comment:
Having a little trouble trying to interpret this plot. So on the right of the plot you have lower regulatization and have more features?

Reply:
>Having a little trouble trying to interpret this plot. So on the right of the plot you have lower regulatization and have more features?

Yup. The right shows low regularization and the left shows high regulariztion. The number of features is the blue line and the green is the accuracy on the right y axis. As the regularization lowers the number of features increases and so does the accuracy.